### PR TITLE
perf: reduce CPU overhead in hot read/write paths

### PR DESCRIPTION
**Pull request by ben1009**
## Summary

Four micro-optimizations targeting the hottest code paths in the LSM engine, reducing CPU overhead on reads and writes.

### 1. Bloom hash pass-through
`get()` now computes `hash_key` once and threads it through `lookup_memtable` → `lookup_sst_raw` → `point_get_with_hash`, eliminating redundant ahash calls per L0 SST on cache miss.

### 2. BlockBuilder pre-allocation
`BlockBuilder::new` pre-allocates `Vec::with_capacity(block_size)` for data and `block_size/32` for offsets, eliminating ~12 Vec-doubling reallocations per block built during flush and compaction.

### 3. Block decode from_vec
`Block::decode_from_vec` takes ownership of the `Vec<u8>` from `FileObject::read` and converts to `Bytes` zero-copy, eliminating one allocation + memcpy per uncached block read.

### 4. Block iterator binary search
`BlockIterator::cmp_entry_at` compares the target key directly against raw prefix+suffix bytes, avoiding full key reconstruction (`clear` + 2x `append`) on every binary search probe.

## Benchmark Results

New cold-cache workloads (4-block cache, forces disk reads):

| Workload | Baseline | Optimized | Delta |
|----------|----------|-----------|-------|
| cold_point_get/inline | 355 ns | 323 ns | **-9.0%** |
| cold_point_get/vlog | 613 ns | 564 ns | **-8.0%** |
| flush_throughput/inline | 5.17 ms | 4.65 ms | **-10.1%** |
| cold_scan/inline | 3.70 ms | 3.62 ms | -2.2% |

Existing warm-cache workloads:

| Workload | Baseline | Optimized | Delta |
|----------|----------|-----------|-------|
| read_point_get/vlog_cached | 750 ns | 717 ns | **-4.4%** |
| read_scan/inline | 14.2 ms | 13.4 ms | **-5.3%** |

## Verification
- [x] `cargo fmt --check` passes
- [x] `cargo clippy` passes (no warnings)
- [x] 136 tests pass
- [x] Code review: no bugs or edge case issues found

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive CPU optimization benchmarks with workloads, before/after timings, and analysis of mode-specific gains.
* **Performance / Benchmarks**
  * New benchmarks for point-get latency, full-scan latency, and flush throughput under constrained cache and heavy flush conditions.
* **Reliability & Efficiency**
  * Safer block decoding, fewer data copies, reduced lookup work and allocations, and more efficient iterator seeks.
* **Error messages**
  * More specific error reporting for reads after a file is closed.

### CODE REVIEW

File: kv-engine/src/block.rs [hunk 1]
```diff
@@ -47,6 +47,25 @@ impl Block {
...
+    pub fn decode_from_vec(data: Vec<u8>) -> Self {
+        let num_of_elements = (&data[data.len() - SIZE_OF_U16..]).get_u16() as usize;
+        let offset = data.len() - SIZE_OF_U16 - num_of_elements * SIZE_OF_U16;
+
+        let buf = Bytes::from(data); // zero-copy: Vec → Bytes (SHARED representation)
+        let datas = buf.slice(..offset);
+        let offsets = buf[offset..buf.len() - SIZE_OF_U16]
+            .chunks(SIZE_OF_U16)
+            .map(|mut x| x.get_u16())
+            .collect();
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block.rs:85]: Addressed in 28a85fa. Added `assert!` bounds checks for `data.len() >= SIZE_OF_U16` and `data.len() >= required` before any slicing. We chose `assert!` over returning `Result` to stay consistent with the existing `decode(&[u8])` method, which also panics on truncated input. Both methods now fail early with a descriptive message instead of producing garbage on corrupted data.

File: kv-engine/src/block/iterator.rs [hunk 2]
```diff
@@ -121,16 +121,45 @@ impl BlockIterator {
...
+            ord => return ord,
+        }
+        // If the target is shorter than the overlap length, the entry's
+        // prefix already exceeds the target.
+        if target.len() < overlap_len {
+            return std::cmp::Ordering::Greater;
+        }
+        // Compare the suffix portion
+        suffix.cmp(&target[overlap_len..])
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block/iterator.rs:165]: Addressed in 28a85fa. Added `assert!` bounds checks on entry data length (>= SIZE_OF_U16 * 2) and suffix length (<= remaining data). Also capped `prefix_cmp_len` with `self.first_key.raw_ref().len()` to guard against corrupted `overlap_len` exceeding the first key boundary.

File: kv-engine/src/lsm_storage.rs [hunk 3]
```diff
@@ -676,16 +676,17 @@ impl LsmStorageInner {
...
         let state = self.state.load_full();
+        let bloom_hash = crate::table::bloom::hash_key(key);
         // Check memtable first — route through resolve_vlog_value_bytes for
         // zero-copy inline slicing (refcount bump instead of heap allocation).
-        if let Some((value, kind)) = self.lookup_memtable(&state, key)? {
+        if let Some((value, kind)) = self.lookup_memtable(&state, key, bloom_hash)? {
             return match kind {
                 KvKind::ValuePointer => self.resolve_vlog_value_bytes(key, value.unwrap()),
                 KvKind::Inline => Ok(value),
             };
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/lsm_storage.rs:689]: Addressed in 28a85fa. Replaced bare `.unwrap()` with `.expect("ValuePointer kind must have a value")` at both call sites (lines 684 and 691) to document the invariant.

File: kv-engine/benches/vlog_benchmarks.rs [hunk 4]
```diff
@@ -403,12 +403,220 @@ fn bench_write_amplification(_c: &mut Criterion) {
...
+            let mut i = 0usize;
+            b.iter(|| {
+                let result = lsm.get(&sst_keys[i % sst_keys.len()]).unwrap();
+                black_box(result);
+                i += 1;
+            })
+        });
+
+        lsm.close().unwrap();
+        drop(dir);
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/benches/vlog_benchmarks.rs]: Addressed in 28a85fa. Replaced `lsm.close().unwrap()` with `drop(lsm)` at all three new benchmark teardown sites (cold_point_get, flush_throughput, cold_scan) to avoid potential lock-order deadlocks.

File: kv-engine/src/block.rs [hunk 5]
```diff
@@ -47,6 +47,25 @@ impl Block {
...
+
+        Self {
+            data: datas,
+            offsets,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ben1009) [reply kv-engine/src/block.rs:91]: Addressed in 28a85fa — added `assert!` bounds checks for minimum length and offset table size before any slicing. We chose `assert!` over returning `Result` to stay consistent with the existing `decode(&[u8])` method which also panics on truncated input. Changing just `decode_from_vec` to return `Result` would create an API inconsistency, and the caller (`read_block`) would need to propagate the error through the block cache layer. The `assert!` approach fails fast with a clear message on corrupted data rather than silently producing garbage.
